lib/logdna/client.rb

```diff
@@ -44,95 +50,125 @@ def process_message(msg, opts = {})
-      }
-      Thread.new { start_timer.call }
+    def write_to_buffer(msg, opts)
+      processed_message = process_message(msg, opts)
+      message_size = processed_message.to_s.bytesize
+
+      buffer_byte_size = @lock.synchronize do
+        @buffer.push(Message.new(processed_message, message_size))
+        @buffer_byte_size += message_size
+      end
+
+      if buffer_byte_size >= @flush_threshold
+        @flush_scheduler.unschedule_flush
+        flush
+      else
+        @flush_scheduler.schedule_flush
+      end
     end
 
-    def write_to_buffer(msg, opts)
-      if @lock.try_lock
-        processed_message = process_message(msg, opts)
-        new_message_size = processed_message.to_s.bytesize
-        @buffer.push(processed_message)
-        @buffer_byte_size += new_message_size
-        @flush_scheduled = true
-        @lock.unlock
-
-        if @flush_limit <= @buffer_byte_size
-          flush
+    ##
+    # Asynchronously flush all logs in the buffer
+    def flush
+      Concurrent::Future.execute({ executor: @thread_pool }) { flush_sync }
+    end
+
+    ##
+    # Concurrently flush all logs in the buffer
+    def flush_sync
+      slices = @lock.synchronize do
+        # Minimize work in the critical section; figure out the slice to send and get out
+        if @buffer.any?
+          pop_points = []
+          pop_count = 0
+          running_size = 0
+
+          # Iterate backwards through array to avoid unnecessary memory allocation
+          @buffer.reverse_each.with_index do |message, index|
+            pop_count += 1
+            running_size += message.size
+            if running_size > @flush_limit || index == @buffer.length - 1
+              pop_points.push(pop_count)
+              @buffer_byte_size -= running_size
+              pop_count = 0
+              running_size = 0
+            end
+          end
+
+          slices = []
+          pop_points.each do |pop_point|
+            slices.push(@buffer.pop(pop_point).map { |message| message.source })
+          end
+
+          slices
         else
-          schedule_flush
-        end
-      else
-        @side_message_lock.synchronize do
-          @side_messages.push(process_message(msg, opts))
+          []
         end
       end
-    end
 
-    # This method has to be called with @lock
-    def send_request
-      @side_message_lock.synchronize do
-        @buffer.concat(@side_messages)
-        @side_messages.clear
+      slices.reverse_each do |slice|
+        send_request(slice)
       end
+    end
 
+    def send_request(slice)
+      # TODO: Remove instance-global request object
       @request.body = {
         e: "ls",
-        ls: @buffer
+        ls: slice
       }.to_json
 
-      handle_exception = lambda do |message|
-        @internal_logger.debug(message)
-        @exception_flag = true
-        @side_message_lock.synchronize do
-          @side_messages.concat(@buffer)
-        end
-      end
+      flush_id = Random.uuid
```

Comment:
If we only flush one at a time, would this make more sense if it was based on the time? Or is the timestamp from the logger just fine, here?

Reply:
I don't see a reason to limit the number of sending requests to just one at a time. I just set it up that way to be consistent with previous behavior. I think at some point we can make it user configurable.

The logger does create a timestamp for us